pkg/sif/update_test.go

```diff
@@ -152,3 +153,125 @@ func TestUpdate(t *testing.T) {
 		})
 	}
 }
+
+//nolint:dupl
+func TestAppendImage(t *testing.T) {
+	r := rand.NewSource(randomSeed)
+	newImage, err := random.Image(64, 1, random.WithSource(r))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name string
+		base string
+		opts []sif.AppendOpt
+	}{
+		{
+			name: "Default",
+			base: "hello-world-docker-v2-manifest",
+			opts: []sif.AppendOpt{},
+		},
+		{
+			name: "WithReference", // Replaces many layers with a single layer
+			base: "hello-world-docker-v2-manifest",
+			opts: []sif.AppendOpt{
+				sif.OptAppendReference(name.MustParseReference("myimage:v1", name.WithDefaultRegistry(""))),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			sifPath := corpus.SIF(t, tt.base, sif.OptWriteWithSpareDescriptorCapacity(8))
+			fi, err := ssif.LoadContainerFromPath(sifPath)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ofi, err := sif.FromFileImage(fi)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if err := ofi.AppendImage(newImage, tt.opts...); err != nil {
+				t.Fatal(err)
+			}
+
+			if err := fi.UnloadContainer(); err != nil {
+				t.Fatal(err)
+			}
+
+			b, err := os.ReadFile(sifPath)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			g := goldie.New(t,
+				goldie.WithTestNameForDir(true),
+			)
+
+			g.Assert(t, tt.name, b)
+		})
+	}
+}
+
+//nolint:dupl
+func TestAppendIndex(t *testing.T) {
+	r := rand.NewSource(randomSeed)
+	newIndex, err := random.Index(64, 1, 1, random.WithSource(r))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name string
+		base string
+		opts []sif.AppendOpt
+	}{
+		{
+			name: "Default",
+			base: "hello-world-docker-v2-manifest",
+			opts: []sif.AppendOpt{},
+		},
+		{
+			name: "WithReference", // Replaces many layers with a single layer
+			base: "hello-world-docker-v2-manifest",
+			opts: []sif.AppendOpt{
+				sif.OptAppendReference(name.MustParseReference("myindex:v1", name.WithDefaultRegistry(""))),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			sifPath := corpus.SIF(t, tt.base, sif.OptWriteWithSpareDescriptorCapacity(8))
+			fi, err := ssif.LoadContainerFromPath(sifPath)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ofi, err := sif.FromFileImage(fi)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if err := ofi.AppendIndex(newIndex, tt.opts...); err != nil {
+				t.Fatal(err)
+			}
+
+			if err := fi.UnloadContainer(); err != nil {
+				t.Fatal(err)
+			}
+
+			b, err := os.ReadFile(sifPath)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			g := goldie.New(t,
+				goldie.WithTestNameForDir(true),
+			)
+
+			g.Assert(t, tt.name, b)
+		})
+	}
+}
```

Comment:
In the event that the same reference is applied to two different `AppendImage`/`AppendIndex` calls, I wonder what the expected behaviour should be? At present, if I run this test:

```suggestion
}

func TestAppendMultiple(t *testing.T) {
	r := rand.NewSource(randomSeed)
	image1, err := random.Image(64, 1, random.WithSource(r))
	if err != nil {
		t.Fatal(err)
	}

	image2, err := random.Image(64, 1, random.WithSource(r))
	if err != nil {
		t.Fatal(err)
	}

	sifPath := corpus.SIF(t, "hello-world-docker-v2-manifest", sif.OptWriteWithSpareDescriptorCapacity(8))
	fi, err := ssif.LoadContainerFromPath(sifPath)
	if err != nil {
		t.Fatal(err)
	}

	ofi, err := sif.FromFileImage(fi)
	if err != nil {
		t.Fatal(err)
	}

	referenceOpt := sif.OptAppendReference(name.MustParseReference("myimage:v1", name.WithDefaultRegistry("")))

	if err := ofi.AppendImage(image1, referenceOpt); err != nil {
		t.Fatal(err)
	}
	if err := ofi.AppendImage(image2, referenceOpt); err != nil {
		t.Fatal(err)
	}

	if err := fi.UnloadContainer(); err != nil {
		t.Fatal(err)
	}

	b, err := os.ReadFile(sifPath)
	if err != nil {
		t.Fatal(err)
	}

	g := goldie.New(t,
		goldie.WithTestNameForDir(true),
	)

	g.Assert(t, "image", b)
}
```

I end up with two images with the same reference:

```
$ panoplia manifest --local pkg/sif/testdata/TestAppendMultiple/image.golden | jq
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 525,
      "digest": "sha256:432f982638b3aefab73cc58ab28f5c16e96fdb504e8c134fc58dff4bae8bf338",
      "platform": {
        "architecture": "arm64",
        "os": "linux",
        "variant": "v8"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 423,
      "digest": "sha256:e4ccb8ccf3c6e47f195a31b13bcf7f3a0075d06dabe450d373f2720f72bab898",
      "annotations": {
        "org.opencontainers.image.ref.name": "myimage:v1"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 423,
      "digest": "sha256:e4b10cf7f46deb943dfcd738afa9bccebab0e5f765fa1464622f66952ea0bee4",
      "annotations": {
        "org.opencontainers.image.ref.name": "myimage:v1"
      }
    }
  ]
}
```

If this were a registry, I guess you'd end up with the last image being the only one with `myindex:v1`. Options that come to mind for me:

1. Keep them both with the tag... this will likely throw an error when the user tries to access `myindex:v1`.
1. Remove the tag from the old image, migrate it to the new one. Keep the old image, but without the tag.
1. Replace the old image with the new one. This might be surprising, given the the name `AddImage`.
1. Throw an error and refuse to `AddImage` with a duplicate tag.

Think I'd vote for emulating the registry behaviour, since that's probably the least surprising for someone in the containers ecosystem. Thoughts?

Reply:
Yeah, that makes sense I think, to emulate the registry behaviour.

Reply:
See - https://github.com/sylabs/oci-tools/pull/87/commits/ff2ac3c1a24ef1e6bfb49ed34094bd29b4feccfc - not sure if you can suggest a more elegant way to do it?

Reply:
After a long think and some experimentation... think I've come up with a decent way to unset the reference without having to define a new annotation just to get the GGCR code to do what you want it to.

Pushed up a commit here with it: https://github.com/tri-adam/oci-tools/commit/c1bda85bd6f330b00fb58e5c2c1e21f43e1a7ca6. Could you take a look and let me know if that looks cleaner? Basically just editing the descriptors in the index manifest directly via wrapping...